### PR TITLE
Added Istio support to KubernetesPodOperator

### PR DESCRIPTION
**Pull request by joshuayeung**
This PR added IstioKubernetesPodOperator and IstioPodManager for making KubernetesPodOperator works in Kubernetes with Istio enabled.

Istio will inject a sidecar called istio-proxy inside your pod and handle the traffic for you. But this hinders the standard workflow of the Kubernetes executor and Kubernetes Pod Operator. It is because both of them use the Pod Phase to determine the status of the dag task. Airflow will consider that the task is finished when the Pod Phase is successful and failed. Because of the istio-proxy sidecar, the pod will always be running. Just the task container itself is completed. This will lead to the Pod Phase become Not Ready in the success task, and Error in the failed task. In Airflow UI, you will see the task always running and never-ending.

The best way is to implement an Istio Kubernetes Pod Operator that shut down the pod when the base container (the container that runs your task) is completed. It involves creating an Istio pod launcher.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by sfc-gh-atatara** [comment]
Thank you for that change, our initial test proved it working

**Comment by ashb** [comment]
I don't think this should be a new operator!

**Comment by potiuk** [comment]
> I don't think this should be a new operator!

Agree. I think it should be merged into KPO (istio=True). It's logic might be separated into classes but it should be one module and operator to use by the users.

**Comment by dstandish** [comment]
> I don't think this should be a new operator!

was just gonna say the same thnig

**Comment by sfc-gh-atatara** [comment]
it is actioned as single operator now

File: airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py [hunk 1]
```diff
@@ -453,7 +455,7 @@ def cleanup(self, pod: k8s.V1Pod, remote_pod: k8s.V1Pod):
         if not self.is_delete_operator_pod:
             with _suppress(Exception):
                 self.patch_already_checked(remote_pod)
-        if pod_phase != PodPhase.SUCCEEDED:
+        if (not self.istio_enabled and pod_phase != PodPhase.SUCCEEDED) or (self.istio_enabled and pod_phase != PodPhase.SUCCEEDED and not self.pod_manager.container_is_succeeded(pod, 'base')):
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:458]: you will need to enable pre-commit locally see https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst

Comment (dstandish) [inline airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:458]: this  expression can probably be simplified. i would try to do so.  it's too hard to understand as it is.

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 2]
```diff
@@ -84,6 +84,36 @@ def container_is_running(pod: V1Pod, container_name: str) -> bool:
...
+    """
+    if not container_is_completed(pod, container_name):
+        return False
+    container_statuses = pod.status.container_statuses if pod and pod.status else None
+    if not container_statuses:
+        return False
+    container_status = next(iter([x for x in container_statuses if x.name == container_name]), None)
+    if not container_status:
+        return False
+    return container_status.state.terminated.exit_code == 0
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py:114]: seems you are duplicating a lot of logic here.  looks like you could pull out a `get_container_status` function

File: airflow/providers/cncf/kubernetes/utils/pod_manager.py [hunk 3]
```diff
@@ -272,6 +304,10 @@ def await_pod_completion(self, pod: V1Pod) -> V1Pod:
             remote_pod = self.read_pod(pod)
             if remote_pod.status.phase in PodPhase.terminal_states:
                 break
+            if self.istio_enabled and remote_pod.status.phase == PodPhase.RUNNING and self.container_is_completed(
+                remote_pod, 'base'
```
Comment (dstandish) [inline airflow/providers/cncf/kubernetes/utils/pod_manager.py:308]: `'base'` is defined in a constant somewhere

**Comment by eladkal** [comment]
@joshuayeung are you still working on this PR?

**Comment by Owen-CH-Leung** [comment]
any chances to reopen this ? I know @joshuayeung personally and have his consent to continue to work on this PR. Thanks

**Comment by sfc-gh-atatara** [comment]
@Owen-CH-Leung I am actually interested in this work too. I have worked on this in my company and we are seeing stable results in development environment.

**Comment by dstandish** [comment]
i reopened it... but you might have to create a new PR anyway, unless joshua grants you permission to his fork

**Comment by eladkal** [comment]
for anyone who wants to continue the work started on this PR, it's much better to cherry pick the commits to your own fork/branch and continue from there. By doing so you are working independently and also save the credit for the original author of the commits that you are building on top of.

**Comment by Owen-CH-Leung** [comment]
Ok I'll cherry pick @joshuayeung 's commits and start a new PR from there
@sfc-gh-atatara Yeah Joshua and I are colleagues before and we actually implement this solution and have it running in our environment for over a year, so we can take care of this =)

**Comment by eladkal** [comment]
Closing infavor of https://github.com/apache/airflow/pull/31389

**Pull request closed without merging**